### ISSUE TITLE
[modbus] Refactor things into channels similar to mqtt.generic

### ISSUE THREAD

**Issue by ssalonen** [openhab/openhab-addons#14058]
openHAB 4.0 is in the works and this is the right time to consider breaking changes. I create this issue to bring visibility and open up discussion. cc @Rossko57 as FYI

### Background
Modbus binding was one of the first bindings with complex thing hierarchy...and that shows. At the time, I decided to go for the following thing hierarchy

Read/write `data` things:
```
endpoint (tcp or serial)
-> poller (reading block of data, every x milliseconds)
---> data (referring to one piece of data from poller, optionally how to write data)
```

Write-only `data` things:
```
endpoint (tcp or serial)
--> data (how to write data)
```

This setup introduces lot of things which is cumbersome to edit in MainUI. For example, MainUI YAML view is optimized to edit single thing, not multiple things. Often with Modbus, many of the `data` things are very similar, just the "address" is different.

There's other negative downsides with several things, for example a lot of thing status spamming in logs when poller goes OFFLINE.

### Proposal (WIP)

We can take example of mqtt.generic binding where one can configure channels.

The exact same approach is applicable to modbus. Instead of data things, we could have configurable channels for `poller` (read/write behaviour) and `tcp`/`serial` (write-only behaviour). Channels could be specialized for the item type, allowing ease of use, for example

`Switch` channel could have `on_value` and `off_value` to encode/decode raw bit pattern from/to modbus.

We could allow transformations for `writeStart` and `readStart` addresses for more dynamic behaviour but without complexity of `JSON` format, see e.g. Rollershutter example https://www.openhab.org/addons/bindings/modbus/#rollershutter.

### Some considerations

1. Many modbus users might have tens if not hundreds of `data` things. We probably should consider if it is possible to create a tool to assist with migration? This might get complicated though.
2. We need to ensure no breakages with `modbux.xxx` device bindings. I think this is not an issue, they do not rely on `poller` and `data` things.

**Comment by lsiepel** [comment]
As this refactoring mainly focuses on thing structure if i understand it well, i'm wondering if these two 
 enahncements are also considered?
https://github.com/openhab/openhab-addons/issues/9576
https://github.com/openhab/openhab-addons/issues/13247

**Comment by ssalonen** [comment]
Yes, returning raw data as hex string or RawType is alreeady in the plans

**Comment by ssalonen** [comment]
UI Prototype 

![image](https://user-images.githubusercontent.com/315866/209853487-fb84ede8-338f-4269-90e8-e7695085e3f8.png)
![image](https://user-images.githubusercontent.com/315866/209853514-f4c93a89-da5c-4df2-9a24-840af82e8e91.png)

**Comment by ssalonen** [comment]
This is not going to make OH 4.0, for sure. I have had not have time to work on this during the last months, really. 

Changes like this require more community feedback, and perhaps even guidance on the thing/channel design.

I have cleaned up the the initial description, documenting the current state of the prototype.

**Comment by J-N-K** [comment]
Please also have a look at https://github.com/openhab/openhab-core/pull/3355

**Comment by ssalonen** [comment]
I am thinking of working around https://github.com/openhab/openhab-webui/issues/1478 by introducing multiple channel types, one for each Item type (yikes!)

Example:
![image](https://github.com/openhab/openhab-addons/assets/315866/b9b015f0-abe6-4204-a010-29aa9620b5f2)

I probably need to introduce even more channel types, for example "Interpret read data as percent (0-100%)" for all item types. One could for example, first scale the raw modbus data to % and then transform it to Switch (ON/OFF) using channel transform. 

@ghys @Rossko57 this is exploding pretty rapidly... Would it be ok to simply remove the item type filtering when linking channels? https://github.com/openhab/openhab-webui/issues/1478#issuecomment-1437361133

**Comment by ssalonen** [comment]
I guess one alternative to making the linking more "flexible" in UI would be to change the channel structure more to the original binding, channel names mapping to the item names

- readTo**Number**
- readTo**Dimmer**
- readTo**Rollershutter**
- ...
- writeCoilFrom**Number**
- writeCoilFrom**Dimmer**
- ...
- writeRegisterFrom**Number**
- writeRegisterFrom**Dimmer**
- ...

The functionality of interpreting the binary data to openHAB data types (e.g. converting to ON/OFF, converting to PercentType) from modbus would be facilitated by **transformations provided by the binding**. Transformations could be combined with ∩.

This is almost analogous to the original binding channels (see https://www.openhab.org/addons/bindings/modbus/#data-thing), but with read and writes separated, as well as coil writes and register writes separated. Also, I would like to have the configuration at the channel level, no need to introduce additional things (the `data` thing in the original binding). It should be possible to instantiate multiple channels of same channel type, but just with different configuration.

This separation allows a bit cleaner UI, not showing irrelevant options that resulted in channel configuration error in the original binding (see the complexity of the validation logic here: https://github.com/openhab/openhab-addons/blob/a31b1578be628f0de10171167cf66d91cb4953ff/bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusDataThingHandler.java#L500-L784). For example,
- there is no point specifying value type with coil writes, the data type in modbus is a single bit in the typical case
- it is reasonable to expect read/write items having different read configuration (e.g. read current room temperature) and write configuration (e.g. _command_ target room temperature). [Separating these "concerns"](https://en.wikipedia.org/wiki/Separation_of_concerns) (as they are completely different operations in modbus) is perhaps more appropriate for a low level generic binding like this. By combining these channels is still possible to build logical items.

**Comment by lsiepel** [comment]
It’s been a while, @ssalonen, so apologies for bluntly tagging you here! I was just wondering whether this work is still under development or consideration.

**Comment by ssalonen** [comment]
Hi @lsiepel ! No, this is no longer under work.

The design did not survive the core changes introduced some major versions back, as explained in https://community.openhab.org/t/is-it-possible-to-create-a-modbus-bundle-to-avoid-having-loads-of-individual-things/161711/13 

As far as I could gather, the architecture made it hard to support this type of generic binding design. Of course, this was ages ago, and perhaps it is different now

**Comment by ssalonen** [comment]
The main obstacle I encountered was that WebUI and core both were against the idea of linking channel to item when the types differ. 

I also encountered surprising issues linking non-UoM Number channel with quantity-aware Number item (was it so that openhab core filtered out the commands since some version)

Target I had was to streamline both with heavy use of rules - this is "standard" use case with generic modbus entities
